src/libraries/System.Linq/tests/CountTests.cs

```diff
@@ -126,5 +126,86 @@ public void NullPredicate_ThrowsArgumentNullException()
             Func<int, bool> predicate = null;
             AssertExtensions.Throws<ArgumentNullException>("predicate", () => Enumerable.Range(0, 3).Count(predicate));
         }
+
+        [Fact]
+        public void NonEnumeratingCount_NullSource_ThrowsArgumentNullException()
+        {
+            AssertExtensions.Throws<ArgumentNullException>("source", () => ((IEnumerable<int>)null).TryGetNonEnumeratingCount(out _));
+        }
+
+        [Theory]
+        [MemberData(nameof(NonEnumeratingCount_SupportedEnumerables))]
+        public void NonEnumeratingCount_SupportedEnumerables_ShouldReturnExpectedCount<T>(int expectedCount, IEnumerable<T> source)
+        {
+            Assert.True(source.TryGetNonEnumeratingCount(out int actualCount));
+            Assert.Equal(expectedCount, actualCount);
+        }
+
+        [Theory]
+        [MemberData(nameof(NonEnumeratingCount_UnsupportedEnumerables))]
+        public void NonEnumeratingCount_UnsupportedEnumerables_ShouldReturnFalse<T>(IEnumerable<T> source)
+        {
+            Assert.False(source.TryGetNonEnumeratingCount(out int actualCount));
+            Assert.Equal(-1, actualCount);
+        }
+
+        [Fact]
+        public void NonEnumeratingCount_ShouldNotEnumerateSource()
+        {
+            bool isEnumerated = false;
+            Assert.False(Source().TryGetNonEnumeratingCount(out int count));
+            Assert.Equal(-1, count);
+            Assert.False(isEnumerated);
+
+            IEnumerable<int> Source()
+            {
+                isEnumerated = true;
+                yield return 42;
+            }
+        }
+
+        public static IEnumerable<object[]> NonEnumeratingCount_SupportedEnumerables()
+        {
+            yield return WrapArgs(4, new int[]{ 1, 2, 3, 4 });
+            yield return WrapArgs(4, new List<int>(new int[] { 1, 2, 3, 4 }));
+            yield return WrapArgs(4, new Stack<int>(new int[] { 1, 2, 3, 4 }));
+
+            yield return WrapArgs(0, Enumerable.Empty<string>());
+#if LINQ_OPTIMIZED_FOR_SPEED
```

Comment:
We have several tests conditioned by platform to account for this.  If we're going to introduce this ifdef model, we should use it in those previous cases, too.

Reply:
Ah, interesting, I hadn't realized we had a runtime check. Fixed.